File: Cpp/2241_design_an_atm_machine.cpp
// https://leetcode.cn/problems/design-an-atm-machine/

class ATM {
public:
    ATM() {
        valids_ = {20, 50, 100, 200, 500};
        cnts_ = {0, 0, 0, 0, 0};
    }
    
    void deposit(vector<int> banknotesCount) {
        for (int i = 0; i < 5; ++i) {
            cnts_[i] += banknotesCount[i];
        }
    }
    
    vector<int> withdraw(int amount) {
        vector<int> res(5, 0);
        for (int i = 4; i >= 0; --i) {
            res[i] = min(cnts_[i], amount / valids_[i]);
            amount -= res[i] * valids_[i];
        }
        if (amount != 0) return vector<int>{-1};
        for (int i = 0; i < 5; ++i) {
            cnts_[i] -= res[i];
        }
        return res;
    }

    vector<int> valids_, cnts_;
};

/**
 * Your ATM object will be instantiated and called as such:
 * ATM* obj = new ATM();
 * obj->deposit(banknotesCount);
 * vector<int> param_2 = obj->withdraw(amount);
 */

File: Python3/2241_design_an_atm_machine.py
# https://leetcode.cn/problems/design-an-atm-machine/

class ATM:

    def __init__(self):
        self.valids = [20, 50, 100, 200, 500]
        self.cnts = [0 for _ in range(5)]

    def deposit(self, banknotesCount: List[int]) -> None:
        for i, cnt in enumerate(banknotesCount):
            self.cnts[i] += cnt

    def withdraw(self, amount: int) -> List[int]:
        res = [0 for _ in range(5)]
        for i in range(4, -1, -1):
            k = min(self.cnts[i], amount//self.valids[i])
            amount -= k * self.valids[i]
            res[i] = k
        
        if amount != 0:
            return [-1]
        
        for i, cnt in enumerate(res):
            self.cnts[i] -= cnt
        
        return res


# Your ATM object will be instantiated and called as such:
# obj = ATM()
# obj.deposit(banknotesCount)
# param_2 = obj.withdraw(amount)